### PR TITLE
Add Information Disclosure and use only secure HostKeys

### PR DESCRIPTION
**Pull request by CookieCr2nk**
Hello,

I added some improvments to the SSH Config.

1) Prevent information disclosure

2) Use only secure Hostkeys

regards,

CookieCr2nk

### CODE REVIEW

**Comment by k4yt3x** [comment]
Thanks for the PR. We'll definitely have to adjust the wording here a bit, and I'll validate the specific configs.

CC @IceCodeNew

Supplemental information for reference:

- https://manpages.debian.org/buster/openssh-server/sshd_config.5.en.html#DebianBanner

**Comment by IceCodeNew** [comment]
Sorry for the delay, just came back from a vacation. Working on it...

File: sshd_config
Comment (IceCodeNew) [inline sshd_config]:
```suggestion
```

Comment (IceCodeNew) [reply sshd_config]: It should be fine only to enable the ssh_host_ed25519_key

Refer to: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/configuring_basic_system_settings/assembly_using-secure-communications-between-two-systems-with-openssh_configuring-basic-system-settings#making-openssh-more-secure_assembly_using-secure-communications-between-two-systems-with-openssh

**Review by IceCodeNew — APPROVED** [review]

**Comment by k4yt3x** [comment]
I typed a bunch of things, added screenshots and I just forgot to send it and they've all be cleared after reboot. Darn.

Anyhow, we can add `DebianBanner`, but we'll need to add it commented out since it's a Debian-specific option and sshd refuses to start with unknown options:

image

As for `HostKey`, I don't agree that we should disable all host keys. While we can arguably disable ECDSA since it always uses NIST P-curves and is, therefore, not ideal, I don't see why we can't use RSA. According to the Ed25519 website:

> ...breaking it has similar difficulty to breaking NIST P-256, RSA with ~3000-bit keys, strong 128-bit block ciphers, etc.

...therefore, 3072-bit RSA or better should be on par with Ed25519 in terms of security. I would advocate for leaving it enabled in the config and disable only ECDSA.

@IceCodeNew

**Comment by k4yt3x** [comment]
I pushed a new commit. Waiting for review.

File: sshd_config
Comment (IceCodeNew) [inline sshd_config]: I believe it is not necessary to include the rsa_key. Here is the reason:
1. The strength of the rsa_key may vary
2. Providing only the ed25519_key should not harm the compatibilities of this project. Thus, I prefer to reduce the attack surface
3. Time to create an SSH connection may also benefit by avoiding the rsa_key since it requires more bits to be sent

P.S. Point 3 is really not a good point, but the tune of this project is quite like that ;-)

File: sshd_config
Comment (IceCodeNew) [inline sshd_config:126]: It might be better to put this option in another config file, and introduce it from /etc/ssh/sshd_config.d/01-debianbanner.conf

Otherwise updating the main config would require users to resolve conflicts manually.

**Review by IceCodeNew — CHANGES_REQUESTED** [review]

File: sshd_config
Comment (IceCodeNew) [reply sshd_config]: Just in case you have missed that, here is the suggestion from the Redhat Inc.
Refer to: [access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/configuring_basic_system_settings/assembly_using-secure-communications-between-two-systems-with-openssh_configuring-basic-system-settings#making-openssh-more-secure_assembly_using-secure-communications-between-two-systems-with-openssh](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/configuring_basic_system_settings/assembly_using-secure-communications-between-two-systems-with-openssh_configuring-basic-system-settings#making-openssh-more-secure_assembly_using-secure-communications-between-two-systems-with-openssh)

**Comment by IceCodeNew** [comment]
kindly ping @k4yt3x

File: sshd_config
Comment (k4yt3x) [reply sshd_config:126]: While I agree with your point, it'll make downloading the config harder. Right now for new servers or even temporary servers I'll just do:

```bash
sudo curl -L k4t.io/sshd -o /etc/ssh/sshd_config
```

If there were to be an additional file more commands will have to be involved. I think we can leave it to the users to put this in `sshd_config.d` if they don't want it to be overwritten over an update.

Comment (IceCodeNew) [reply sshd_config:126]: OK

File: sshd_config
Comment (k4yt3x) [reply sshd_config]: Apologies for the delay. I have just been way too busy recently.

Yeah I made the change after seeing this page. My impression reading the text is that RH is giving an example of how to disable certain host keys, not recommending it to be done like that.

The other thing is, while RSA key can be weaker than ed25519, it can also be equivalently strong (3072 bits) or stronger (4096 bits). It is also required in some cases for FIPS compliance. I am therefore hesitant to outright exclude it.

**Pull request merged**
